File: backend/ai_engine/rag_processor.py
import os
import re
import json

from data_collector import relative_time

GROQ_API_URL = "https://api.groq.com/openai/v1/chat/completions"

class FinancialRAGProcessor:
    """
    Real-data NEPSE assistant.

    Answers are grounded in:
      - Real MeroLagani fundamentals cache (P/E, ROE, EPS, book value, dividend yield, market cap)
      - Real NEPSE live-market snapshot (index, turnover, top gainers/losers, LTP)
      - Real per-scrip news sentiment cache
      - Real technical signals computed from actual price history (transparent scorer)

    No fabricated figures. When a Groq API key is supplied the LLM summarizes the
    same real context; otherwise a deterministic markdown answer is produced.
    """
    def __init__(self):
        self.base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
        self.raw_news_dir = os.path.join(self.base_dir, "data", "raw", "news")
        self.fund_cache_file = os.path.join(self.base_dir, "data", "raw", "fundamentals", "merolagani_cache.json")
        self.snapshot_file = os.path.join(self.base_dir, "data", "nepse_snapshot.json")
        self._fund_cache = None
        self._snapshot_cache = None

    # ── Real data loaders ────────────────────────────────────────────────

    def _load_fundamentals(self):
        if self._fund_cache is None:
            try:
                with open(self.fund_cache_file, encoding='utf-8') as f:
                    self._fund_cache = json.load(f)
            except Exception:
                self._fund_cache = {}
        return self._fund_cache

    def _load_snapshot(self):
        if self._snapshot_cache is None:
            try:
                with open(self.snapshot_file, encoding='utf-8') as f:
                    self._snapshot_cache = json.load(f)
            except Exception:
                self._snapshot_cache = {}
        return self._snapshot_cache

    def _news_for_symbol(self, symbol):
        try:
            with open(os.path.join(self.raw_news_dir, f"{symbol}_news.json"), encoding='utf-8') as f:
                items = json.load(f)
                return items if isinstance(items, list) else []
        except Exception:
            return []

    def _detect_symbol(self, query):
        funds = self._load_fundamentals()
        q = query.upper()
        # Match ticker symbols (longest first so e.g. CBL over CBL... is stable)
        for sym in sorted(funds.keys(), key=len, reverse=True):
            if re.search(rf'\b{re.escape(sym)}\b', q):
                return sym
        # Match full company names
        for sym, f in funds.items():
            name = (f.get('name') or '').upper()
            if name and re.search(rf'\b{re.escape(name)}\b', q):
                return sym
        return None

    def _technical_signal(self, symbol, df_raw):
        """Real transparent scorer output for one symbol from its actual price history."""
        try:
            from indicators import compute_all_indicators
            from ml_model import classifier
            if df_raw is None:
                return None
            sub = df_raw[df_raw['symbol'].str.upper() == symbol]
            if len(sub) < 5:
                return None
            di = compute_all_indicators(sub.sort_values('date'))
            pred = classifier.predict_movement_probabilities(di, cache_key=symbol)
            latest = di.iloc[-1]
            sma_ratio = float(latest.get('close', 0)) / float(latest.get('sma_20', latest.get('close', 1))) if latest.get('sma_20') else None
            return {
                "signal": pred['signal'],
                "confidence": pred['confidenceScore'],
                "price": round(float(latest['close']), 2),
                "rsi": round(float(latest.get('rsi', 50.0)), 2),
                "sma20Ratio": round(sma_ratio, 3) if sma_ratio else None
            }
        except Exception:
            return None

    # ── Real top picks (no hardcoded numbers) ────────────────────────────

    def _build_top_picks(self, limit=5):
        funds = self._load_fundamentals()
        try:
            from data_collector import fetch_price_history_csv
            df_raw = fetch_price_history_csv()
        except Exception:
            df_raw = None

        picks = []
        for sym, f in funds.items():
            pe = f.get('peRatio')
            roe = f.get('roe')
            dy = f.get('dividendYield')
            # Require sane, real valuations: skip loss-making and absurd P/E
            if pe is None or pe <= 0 or pe > 30:
                continue
            if roe is None or roe <= 0:
                continue

            tech = self._technical_signal(sym, df_raw)
            signal = tech['signal'] if tech else 'NEUTRAL'
            confidence = tech['confidence'] if tech else 55.0
            price = tech['price'] if tech and tech['price'] else f.get('price')

            score = 0.0
            parts = []
            if pe < 15:
                score += 2.0
            elif pe < 20:
                score += 1.2
            elif pe < 25:
                score += 0.6
            if roe > 18:
                score += 1.5
            elif roe > 12:
                score += 1.0
            elif roe > 8:
                score += 0.5
            if dy and dy > 4:
                score += 1.5
            elif dy and dy > 2.5:
                score += 1.0
            elif dy and dy > 1.5:
                score += 0.5
            if signal == 'BULLISH':
                score += 1.5
            elif signal == 'NEUTRAL':
                score += 0.4
            else:
                score -= 1.0
            if tech and tech.get('sma20Ratio'):
                score += 0.8 if tech['sma20Ratio'] >= 1.0 else -0.5

            reason_bits = [f"P/E {pe}", f"ROE {roe}%"]
            if dy:
                reason_bits.append(f"div yield {dy}%")
            if tech:
                reason_bits.append(f"{signal} technicals (RSI {tech['rsi']})")
            parts = reason_bits

            picks.append({
                "symbol": sym,
                "name": f.get('name') or sym,
                "sector": f.get('sector') or 'Equity',
                "signal": 'BUY' if signal == 'BULLISH' else ('HOLD' if signal == 'NEUTRAL' else 'SELL'),
                "confidence": round(float(confidence), 1),
                "price": round(float(price), 2) if price else None,
                "reason": "; ".join(parts),
                "score": score
            })

        picks.sort(key=lambda p: p['score'], reverse=True)
        return picks[:limit]

    # ── Context builders ─────────────────────────────────────────────────

    def _market_context(self):
        snap = self._load_snapshot()
        lines = []
        as_of = snap.get('asOf') or ''
        # NEPSE main index from the indices list
        nepse = next((i for i in (snap.get('indices') or []) if i.get('index') == 'NEPSE Index'), None)
        if nepse:
            val = nepse.get('currentValue') or nepse.get('close')
            chg = nepse.get('change')
            pct = nepse.get('perChange')
            turnover = ''
            for item in (snap.get('summary') or []):
                if 'turnover' in (item.get('detail') or '').lower() and item.get('value'):
                    turnover = f", turnover Rs {item['value']:,.0f}"
            lines.append(
                f"Market snapshot (as of {as_of or 'last session'}): NEPSE at {val} "
                f"({chg:+} pts, {pct:+}%){turnover}"
            )
        else:
            lines.append("Market snapshot: live index data currently unavailable.")
        gainers = (snap.get('gainers') or [])[:3]
        losers = (snap.get('losers') or [])[:3]
        if gainers:
            lines.append("Top gainers today: " + "; ".join(
                f"{g.get('symbol')} {g.get('ltp')} ({g.get('percentageChange')}%)" for g in gainers))
        if losers:
            lines.append("Top losers today: " + "; ".join(
                f"{l.get('symbol')} {l.get('ltp')} ({l.get('percentageChange')}%)" for l in losers))
        return "\n".join(lines)

    def _symbol_context(self, symbol):
        funds = self._load_fundamentals()
        f = funds.get(symbol, {})
        tech = self._technical_signal(symbol, None)
        try:
            from data_collector import fetch_price_history_csv
            df_raw = fetch_price_history_csv()
            tech = tech or self._technical_signal(symbol, df_raw)
        except Exception:
            pass

        lines = [f"Company: {symbol} ({f.get('name') or symbol}) - {f.get('sector') or 'Equity'}"]
        if f.get('peRatio') is not None:
            lines.append(f"P/E ratio {f['peRatio']}x, P/B ratio {f.get('pbRatio')}x")
        if f.get('eps') is not None:
            lines.append(f"EPS Rs {f['eps']}, ROE {f.get('roe')}%, book value Rs {f.get('bookValue')}")
        if f.get('dividendYield') is not None:
            lines.append(f"Dividend yield {f['dividendYield']}%")
        if f.get('marketCap') is not None:
            lines.append(f"Market cap Rs {f['marketCap']:,.0f}")
        if f.get('price') is not None:
            lines.append(f"Last traded price Rs {f['price']}")
        if tech:
            lines.append(
                f"Technical (computed from real price history): {tech['signal']} with {tech['confidence']}% confidence, "
                f"RSI {tech['rsi']}, price Rs {tech['price']}"
            )
        news = self._news_for_symbol(symbol)
        if news:
            lines.append("Recent news:")
            for item in news[:6]:
                lines.append(
                    f"- {item.get('title')} ({item.get('pubDate')}) [Sentiment: {item.get('sentimentLabel', 'NEUTRAL')}]"
                )
        return "\n".join(lines)

    # ── Intent helpers ───────────────────────────────────────────────────

    def _is_rec_query(self, query):
        return any(w in query.lower() for w in [
            'buy', 'recommend', 'which stock', 'top pick', 'should i buy',
            'best stock', 'invest', 'stocks to', 'which nepse', 'good stock'
        ])

    def _is_greeting(self, query):
        return any(w in query.lower().strip() for w in [
            'hi', 'hello', 'hey', 'namaste', 'good morning', 'good evening',
            'who are you', 'what can you do', 'help'
        ]) and len(query.strip().split()) <= 4

    # ── Main entry ───────────────────────────────────────────────────────

    def query_financial_document(self, query, groq_api_key=None, company_meta=None, symbol=None):
        if not symbol:
            symbol = self._detect_symbol(query)
        symbol = symbol.upper().strip() if symbol else None

        api_key = groq_api_key or os.environ.get("GROQ_API_KEY")
        is_rec = self._is_rec_query(query)
        is_greet = self._is_greeting(query)

        market_ctx = self._market_context()
        picks = self._build_top_picks(limit=5) if is_rec else []

        context_parts = [market_ctx]
        if symbol:
            context_parts.append(self._symbol_context(symbol))
        context_str = "\n\n".join(context_parts)

        picks_txt = ""
        if picks:
            picks_txt = "\n".join(
                f"- {p['symbol']} ({p['name']}): {p['signal']} {p['confidence']}% confidence, "
                f"price Rs {p['price']}. {p['reason']}."
                for p in picks
            )

        answer = None
        recommendations = picks
        groq_used = False

        if api_key:
            try:
                from groq import Groq
                client = Groq(api_key=api_key)

                if is_greet:
                    system_prompt = (
                        "You are a friendly, intelligent NEPSE financial advisor. Reply warmly and briefly, "
                        "then note you can analyze stocks, fundamentals, and news using live NEPSE data."
                    )
                    user_prompt = f"User: {query}"
                elif is_rec:
                    system_prompt = (
                        "You are an expert NEPSE stock advisor. Recommend specific scrips using ONLY the real data provided "
                        "below. State Buy/Hold/Sell signals with the real confidence and numbers given. "
                        "Never invent P/E, ROE, dividend, price, or target figures that are not in the data. "
                        "If data is missing, say so."
                    )
                    user_prompt = (
                        f"User question: {query}\n\nReal market data:\n{market_ctx}\n\n"
                        f"Top candidates computed from real fundamentals + technicals:\n{picks_txt}"
                    )
                else:
                    system_prompt = (
                        "You are a helpful NEPSE financial assistant. Answer using ONLY the real data provided below. "
                        "Never invent financial figures. If you don't have the data to answer, say you don't have it."
                    )
                    user_prompt = f"Real data:\n{context_str}\n\nUser question: {query}"

                completion = client.chat.completions.create(
                    model="llama-3.3-70b-versatile",
                    messages=[
                        {"role": "system", "content": system_prompt},
                        {"role": "user", "content": user_prompt}
                    ],
                    temperature=0.2,
                    max_tokens=600
                )
                answer = completion.choices[0].message.content
                groq_used = True
                print(f"[Groq AI] Response generated (symbol={symbol})")
            except Exception as e:
                print(f"[Groq SDK Error]: {e}")

        if not answer:
            answer = self._fallback_answer(query, symbol, is_rec, is_greet, market_ctx, picks, context_str)

        sources = self._sources_for(symbol, picks)

        return {
            "answer": answer,
            "recommendations": recommendations,
            "citations": [],
            "sources": sources,
            "symbol": symbol or "NEPSE Market",
            "query": query,
            "groqPowered": groq_used
        }

    def _fallback_answer(self, query, symbol, is_rec, is_greet, market_ctx, picks, context_str):
        if is_greet:
            return (
                "Namaste! I can help you with NEPSE stocks using real market data — try:\n"
                "- \"Which stocks should I buy?\"\n"
                "- \"Tell me about NABIL or GBIME\"\n"
                "- \"How did the market do today?\"\n\n"
                "*I answer from live NEPSE data directly.*"
            )
        if is_rec:
            if not picks:
                return (
                    "### Top NEPSE Stock Recommendations\n\n"
                    "I don't have enough real valuation data to rank stocks right now. "
                    "The fundamentals cache may need refreshing.\n\n"
                    f"*Market today:* {market_ctx}"
                )
            lines = ["### Top NEPSE Stock Recommendations (from real NEPSE data)", ""]
            for p in picks:
                lines.append(
                    f"**{p['symbol']}** ({p['name']}) - **{p['signal']}** ({p['confidence']}% confidence)\n"
                    f"- Price: Rs {p['price']} · Reason: {p['reason']}"
                )
            lines.append("")
            lines.append("_" + market_ctx + "_")
            lines.append(
                "\n*Scores are computed from real fundamentals (P/E, ROE, dividend yield) and real price history.*"
            )
            return "\n".join(lines)
        if symbol:
            return (
                f"### {symbol} - Snapshot from Real NEPSE Data\n\n"
                f"{context_str}\n\n"
                "*All figures above are pulled live from NEPSE/MeroLagani data.*"
            )
        return (
            "### NEPSE Market Overview (Real Data)\n\n"
            f"{market_ctx}\n\n"
            "Ask me things like:\n- \"Which stocks should I buy?\"\n- \"Tell me about NABIL\"\n- \"What's the market doing?\"\n\n"
            "*This answer is built from live NEPSE data.*"
        )

    def _sources_for(self, symbol, picks):
        sources = []
        if symbol:
            for item in self._news_for_symbol(symbol)[:6]:
                sources.append({
                    "title": item.get('title', ''),
                    "url": item.get('url', ''),
                    "pubDate": item.get('pubDate', ''),
                    "publishedAgo": relative_time(item.get('pubDate')),
                    "sentimentLabel": item.get('sentimentLabel', 'NEUTRAL')
                })
        if not sources and picks:
            snap = self._load_snapshot()
            sources.append({
                "title": f"NEPSE market snapshot {snap.get('asOf') or ''}".strip(),
                "url": "",
                "pubDate": snap.get('asOf', ''),
                "publishedAgo": "latest session",
                "sentimentLabel": "MARKET"
            })
        return sources


rag_processor = FinancialRAGProcessor()

File: backend/src/ai/ragProcessor.js
// Port of ai_engine/rag_processor.py (FinancialRAGProcessor) to JS.
// Real-data NEPSE assistant: answers are grounded in the same cached
// fundamentals / snapshot / news / price-history files. Uses Groq for the
// LLM summary when GROQ_API_KEY is set, otherwise a deterministic markdown
// answer is built from the same real data.

const dataLoader = require('./dataLoader');
const { computeAllIndicators } = require('./indicators');
const { predictMovementProbabilities } = require('./classifier');

const GROQ_URL = 'https://api.groq.com/openai/v1/chat/completions';
const GROQ_MODEL = 'llama-3.3-70b-versatile';

function _f(v, def) {
  if (v == null) return def != null ? def : 0;
  const x = parseFloat(v);
  return isFinite(x) ? x : (def != null ? def : 0);
}

function _num(v) {
  const x = parseFloat(v);
  return isFinite(x) ? x : 0;
}

function _escapeRegExp(s) {
  return String(s).replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function _detectSymbol(query) {
  const funds = dataLoader.loadFundamentals();
  const q = String(query || '').toUpperCase();
  const keys = Object.keys(funds).sort((a, b) => b.length - a.length);
  for (const sym of keys) {
    if (new RegExp(`\\b${_escapeRegExp(sym)}\\b`).test(q)) return sym;
  }
  for (const [sym, f] of Object.entries(funds)) {
    const name = String(f.name || '').toUpperCase();
    if (name && new RegExp(`\\b${_escapeRegExp(name)}\\b`).test(q)) return sym;
  }
  return null;
}

// Transparent scorer output for one symbol from its actual price history.
function _technicalSignal(symbol) {
  try {
    const hist = dataLoader.loadPriceHistory();
    const rows = hist[symbol];
    if (!rows || rows.length < 5) return null;
    const ind = computeAllIndicators(rows);
    const pred = predictMovementProbabilities(ind);
    const latest = ind[ind.length - 1];
    const sma20 = _f(latest.sma_20, 0);
    const smaRatio = sma20 > 0 ? latest.close / sma20 : null;
    return {
      signal: pred.signal,
      confidence: pred.confidenceScore,
      price: Math.round(latest.close * 100) / 100,
      rsi: Math.round(_f(latest.rsi, 50) * 100) / 100,
      sma20Ratio: smaRatio != null ? Math.round(smaRatio * 1000) / 1000 : null,
    };
  } catch (err) {
    return null;
  }
}

// Real top picks (no hardcoded numbers).
function _buildTopPicks(limit = 5) {
  const funds = dataLoader.loadFundamentals();
  const picks = [];
  for (const [sym, f] of Object.entries(funds)) {
    const pe = _f(f.peRatio, null);
    const roe = _f(f.roe, null);
    const dy = _f(f.dividendYield, null);
    if (pe == null || pe <= 0 || pe > 30) continue;
    if (roe == null || roe <= 0) continue;

    const tech = _technicalSignal(sym);
    const signal = tech ? tech.signal : 'NEUTRAL';
    const confidence = tech ? tech.confidence : 55;
    const price = tech && tech.price ? tech.price : _f(f.price, null);

    let score = 0;
    if (pe < 15) score += 2.0;
    else if (pe < 20) score += 1.2;
    else if (pe < 25) score += 0.6;
    if (roe > 18) score += 1.5;
    else if (roe > 12) score += 1.0;
    else if (roe > 8) score += 0.5;
    if (dy && dy > 4) score += 1.5;
    else if (dy && dy > 2.5) score += 1.0;
    else if (dy && dy > 1.5) score += 0.5;
    if (signal === 'BULLISH') score += 1.5;
    else if (signal === 'NEUTRAL') score += 0.4;
    else score -= 1.0;
    if (tech && tech.sma20Ratio) score += tech.sma20Ratio >= 1 ? 0.8 : -0.5;

    const reasonBits = [`P/E ${pe}`, `ROE ${roe}%`];
    if (dy) reasonBits.push(`div yield ${dy}%`);
    if (tech) reasonBits.push(`${signal} technicals (RSI ${tech.rsi})`);

    picks.push({
      symbol: sym,
      name: f.name || sym,
      sector: f.sector || 'Equity',
      signal: signal === 'BULLISH' ? 'BUY' : signal === 'NEUTRAL' ? 'HOLD' : 'SELL',
      confidence: Math.round(confidence * 10) / 10,
      price: price != null ? Math.round(price * 100) / 100 : null,
      reason: reasonBits.join('; '),
      score,
    });
  }
  picks.sort((a, b) => b.score - a.score);
  return picks.slice(0, limit);
}

function _marketContext() {
  const snap = dataLoader.loadSnapshot();
  const lines = [];
  const asOf = snap.asOf || '';
  const nepse = (snap.indices || []).find(i => i.index === 'NEPSE Index');
  if (nepse) {
    const val = nepse.currentValue != null ? nepse.currentValue : nepse.close;
    const chg = nepse.change != null ? nepse.change : 0;
    const pct = nepse.perChange != null ? nepse.perChange : 0;
    let turnover = '';
    for (const item of snap.summary || []) {
      if (String(item.detail || '').toLowerCase().includes('turnover') && item.value != null) {
        turnover = `, turnover Rs ${Math.round(item.value).toLocaleString('en-IN')}`;
      }
    }
    lines.push(`Market snapshot (as of ${asOf || 'last session'}): NEPSE at ${val} (${chg >= 0 ? '+' : ''}${chg} pts, ${pct >= 0 ? '+' : ''}${pct}%)${turnover}`);
  } else {
    lines.push('Market snapshot: live index data currently unavailable.');
  }
  const gainers = (snap.gainers || []).slice(0, 3);
  const losers = (snap.losers || []).slice(0, 3);
  if (gainers.length) lines.push('Top gainers today: ' + gainers.map(g => `${g.symbol} ${g.ltp} (${g.percentageChange}%)`).join('; '));
  if (losers.length) lines.push('Top losers today: ' + losers.map(l => `${l.symbol} ${l.ltp} (${l.percentageChange}%)`).join('; '));
  return lines.join('\n');
}

function _symbolContext(symbol) {
  const f = dataLoader.loadFundamentals()[symbol] || {};
  const tech = _technicalSignal(symbol);
  const lines = [`Company: ${symbol} (${f.name || symbol}) - ${f.sector || 'Equity'}`];
  if (f.peRatio != null) lines.push(`P/E ratio ${f.peRatio}x, P/B ratio ${f.pbRatio}x`);
  if (f.eps != null) lines.push(`EPS Rs ${f.eps}, ROE ${f.roe}%, book value Rs ${f.bookValue}`);
  if (f.dividendYield != null) lines.push(`Dividend yield ${f.dividendYield}%`);
  if (f.marketCap != null) lines.push(`Market cap Rs ${Math.round(f.marketCap).toLocaleString('en-IN')}`);
  if (f.price != null) lines.push(`Last traded price Rs ${f.price}`);
  if (tech) {
    lines.push(`Technical (computed from real price history): ${tech.signal} with ${tech.confidence}% confidence, RSI ${tech.rsi}, price Rs ${tech.price}`);
  }
  const news = dataLoader.loadNewsForSymbol(symbol);
  if (news && news.length) {
    lines.push('Recent news:');
    for (const item of news.slice(0, 6)) {
      lines.push(`- ${item.title} (${item.pubDate}) [Sentiment: ${item.sentimentLabel || 'NEUTRAL'}]`);
    }
  }
  return lines.join('\n');
}

function _isRecQuery(query) {
  const q = String(query || '').toLowerCase();
  return ['buy', 'recommend', 'which stock', 'top pick', 'should i buy', 'best stock', 'invest', 'stocks to', 'which nepse', 'good stock'].some(w => q.includes(w));
}

function _isGreeting(query) {
  const q = String(query || '').toLowerCase().trim();
  return ['hi', 'hello', 'hey', 'namaste', 'good morning', 'good evening', 'who are you', 'what can you do', 'help'].some(w => q.includes(w)) && q.split(/\s+/).length <= 4;
}

async function _groqComplete(systemPrompt, userPrompt, apiKey) {
  const key = apiKey || process.env.GROQ_API_KEY;
  if (!key) return null;
  try {
    const res = await fetch(GROQ_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${key}` },
      body: JSON.stringify({
        model: GROQ_MODEL,
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: userPrompt },
        ],
        temperature: 0.2,
        max_tokens: 600,
      }),
    });
    if (!res.ok) throw new Error(`Groq HTTP ${res.status}`);
    const data = await res.json();
    return (data.choices && data.choices[0] && data.choices[0].message && data.choices[0].message.content) || null;
  } catch (err) {
    console.log(`[Groq Error] ${err.message}`);
    return null;
  }
}

function _sourcesFor(symbol, picks) {
  const sources = [];
  if (symbol) {
    for (const item of dataLoader.loadNewsForSymbol(symbol).slice(0, 6)) {
      sources.push({
        title: item.title || '',
        url: item.url || '',
        pubDate: item.pubDate || '',
        publishedAgo: dataLoader.relativeTime(item.pubDate),
        sentimentLabel: item.sentimentLabel || 'NEUTRAL',
      });
    }
  }
  if (!sources.length && picks.length) {
    const snap = dataLoader.loadSnapshot();
    sources.push({
      title: `NEPSE market snapshot ${snap.asOf || ''}`.trim(),
      url: '',
      pubDate: snap.asOf || '',
      publishedAgo: 'latest session',
      sentimentLabel: 'MARKET',
    });
  }
  return sources;
}

function _fallbackAnswer(query, symbol, isRec, isGreet, marketCtx, picks, contextStr) {
  if (isGreet) {
    return (
      'Namaste! I can help you with NEPSE stocks using real market data — try:\n' +
      '- "Which stocks should I buy?"\n' +
      '- "Tell me about NABIL or GBIME"\n' +
      '- "How did the market do today?"\n\n' +
      '*I answer from live NEPSE data directly.*'
    );
  }
  if (isRec) {
    if (!picks.length) {
      return (
        '### Top NEPSE Stock Recommendations\n\n' +
        "I don't have enough real valuation data to rank stocks right now. " +
        'The fundamentals cache may need refreshing.\n\n' +
        `*Market today:* ${marketCtx}`
      );
    }
    const lines = ['### Top NEPSE Stock Recommendations (from real NEPSE data)', ''];
    for (const p of picks) {
      lines.push(`**${p.symbol}** (${p.name}) - **${p.signal}** (${p.confidence}% confidence)\n- Price: Rs ${p.price} · Reason: ${p.reason}`);
    }
    lines.push('');
    lines.push(`_${marketCtx}_`);
    lines.push('\n*Scores are computed from real fundamentals (P/E, ROE, dividend yield) and real price history.*');
    return lines.join('\n');
  }
  if (symbol) {
    return `### ${symbol} - Snapshot from Real NEPSE Data\n\n${contextStr}\n\n*All figures above are pulled live from NEPSE/MeroLagani data.*`;
  }
  return (
    '### NEPSE Market Overview (Real Data)\n\n' +
    `${marketCtx}\n\n` +
    'Ask me things like:\n- "Which stocks should I buy?"\n- "Tell me about NABIL"\n- "What\'s the market doing?"\n\n' +
    '*This answer is built from live NEPSE data.*'
  );
}

async function queryFinancialDocument(query, { groqApiKey, symbol } = {}) {
  if (!symbol) symbol = _detectSymbol(query);
  symbol = symbol ? String(symbol).toUpperCase().trim() : null;

  const isRec = _isRecQuery(query);
  const isGreet = _isGreeting(query);

  const marketCtx = _marketContext();
  const picks = isRec ? _buildTopPicks(5) : [];

  const contextParts = [marketCtx];
  if (symbol) contextParts.push(_symbolContext(symbol));
  const contextStr = contextParts.join('\n\n');

  const picksTxt = picks.length
    ? picks.map(p => `- ${p.symbol} (${p.name}): ${p.signal} ${p.confidence}% confidence, price Rs ${p.price}. ${p.reason}.`).join('\n')
    : '';

  let answer = null;
  let groqUsed = false;

  let systemPrompt, userPrompt;
  if (isGreet) {
    systemPrompt = 'You are a friendly, intelligent NEPSE financial advisor. Reply warmly and briefly, then note you can analyze stocks, fundamentals, and news using live NEPSE data.';
    userPrompt = `User: ${query}`;
  } else if (isRec) {
    systemPrompt = 'You are an expert NEPSE stock advisor. Recommend specific scrips using ONLY the real data provided below. State Buy/Hold/Sell signals with the real confidence and numbers given. Never invent P/E, ROE, dividend, price, or target figures that are not in the data. If data is missing, say so.';
    userPrompt = `User question: ${query}\n\nReal market data:\n${marketCtx}\n\nTop candidates computed from real fundamentals + technicals:\n${picksTxt}`;
  } else {
    systemPrompt = 'You are a helpful NEPSE financial assistant. Answer using ONLY the real data provided below. Never invent financial figures. If you don\'t have the data to answer, say you don\'t have it.';
    userPrompt = `Real data:\n${contextStr}\n\nUser question: ${query}`;
  }

  answer = await _groqComplete(systemPrompt, userPrompt, groqApiKey);
  if (answer) {
    groqUsed = true;
    console.log(`[Groq AI] Response generated (symbol=${symbol})`);
  }

  if (!answer) answer = _fallbackAnswer(query, symbol, isRec, isGreet, marketCtx, picks, contextStr);

  return {
    answer,
    recommendations: picks,
    citations: [],
    sources: _sourcesFor(symbol, picks),
    symbol: symbol || 'NEPSE Market',
    query,
    groqPowered: groqUsed,
    engine: 'local-js',
  };
}

module.exports = { queryFinancialDocument };
